zzk/service/hoststate.go

```diff
@@ -119,192 +151,264 @@ func (l *HostStateListener) Spawn(shutdown <-chan interface{}, stateID string) {
-			// wait for the container to exit
-			time := <-containerExit
-			logger.WithField("terminated", time).Debug("Container exited")
-		}
+	// reattach to orphaned container
+	ssdat, containerExit := l.getExistingState(stateID)
+	// TODO: Should we go ahead and pull the service state node from zk and make sure it matches?
 
-		// delete the state from the coordinator
-		t := time.NewTicker(time.Second)
-		defer t.Stop()
-
-		for {
-			if err := DeleteState(l.conn, req); err == client.ErrNoServer {
-				logger.WithError(err).Warn("Server not found, attempting to retry updating service")
-				select {
-				case <-t.C:
-					select {
-					case <-shutdown:
-						return
-					default:
-					}
-				case <-shutdown:
-					return
-				}
-				continue
-			} else if err != nil {
-				logger.WithError(err).Error("Could not delete state for stopped container")
-			}
+	sspth := path.Join("/services", serviceID, stateID)
+	// load the service state node
+	if ssdat == nil {
+		ssdat = &ServiceState{}
+		if err := l.conn.Get(sspth, ssdat); err == client.ErrNoNode {
+			// TODO: mismatched data so shut down container and clean up nodes
+			//  Unless we address the TODO above, there is no container running here
+			return
+		} else if err != nil {
+			logger.WithError(err).Error("Could not load service state")
 			return
 		}
-	}()
-
-	// load the service state node
-	sspth := path.Join("/services", serviceID, stateID)
-	ssdat := &ServiceState{}
-	if err := l.conn.Get(sspth, ssdat); err == client.ErrNoNode {
-		return
-	} else if err != nil {
-		logger.WithError(err).Error("Could not load service state")
-		return
 	}
 
 	done := make(chan struct{})
-	defer func() {
-		close(done)
-	}()
+	defer func() { close(done) }()
 
 	for {
 		// set up a listener on the host state node
 		hspth := l.GetPath(stateID)
 		hsdat := &HostState{}
 		hsevt, err := l.conn.GetW(hspth, hsdat, done)
 		if err == client.ErrNoNode {
-
 			logger.Debug("Host state was removed, exiting")
+			l.cleanUpContainer(stateID)
 			return
 		} else if err != nil {
-
 			logger.WithError(err).Error("Could not watch host state")
 			return
 		}
-		// set up a listener on the service state node
+
+		// set up a listener on the service state node, to ensure the node's existence
 		ok, ssevt, err := l.conn.ExistsW(sspth, done)
 		if err != nil {
 			logger.WithError(err).Error("Could not watch service state")
 			return
-		}
-		if !ok {
+		} else if !ok {
 			logger.Debug("Service state was removed, exiting")
+			l.cleanUpContainer(stateID)
 			return
 		}
 
 		// attach to the container if not already attached
 		if containerExit == nil {
 			containerExit, err = l.handler.AttachContainer(ssdat, serviceID, instanceID)
 			if err != nil {
-
 				logger.WithError(err).Error("Could not attach to container")
+				// TODO: there is a problem with docker here. How do we handle
+				// this behavior?  For now just shut down and clean up nodes
+				l.cleanUpContainer(stateID)
```

Comment:
I think it is correct to shut down and clean up in event of a problem with docker.